design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,594 @@
+   for a namespace called `monitoring` on S3 storage bucket called
+   `test-bucket` could be created at the location
+   `s3://test-bucket/<UUID of monitoring namespace>/repo/`.
+- Accessing the repository requires the storage location and credential
+   information similar to a Kanister Profile CR and a unique password used by
+   Kopia during encryption, along with a unique path prefix mentioned above.
+   [See](https://kopia.io/docs/features/#end-to-end-zero-knowledge-encryption).
+- In the first iteration, users will be required to provide the location and
+   repository information to the controller during the creation of the
+   repository server in the form of Kubernetes Secrets. Future iterations will
+   allow users to use a Key Management Service of choice.
+
+### Kopia Repository Server
+
+- A Kopia Repository Server allows Kopia clients proxy access to the backend
+   storage location through it.
+- At any time, a repository server can only connect to a single repository. Due
+   to this a separate instance of the server will be used for each repository.
+- In Kanister, the server will comprise a K8s `Pod`, `Service` and
+   a `NetworkPolicy`.
+- The pod will execute the Kopia server process exposed to the application via
+   the K8s service and the network policy.
+- Accessing the server requires the service name, a server username, and a
+   password without any knowledge of the backend storage location.
+- To authorize access, a list of server usernames and passwords must be added
+   prior to starting the server.
+- The server also uses TLS certificates to secure incoming connections to it.
+- Kanister users can configure a repository server via a newly added Custom
+   Resource called `RepositoryServer` as described in the following section.
+
+#### Custom Resource Definition
+
+This design proposes a new Custom Resource Definition (CRD) named
+`RepositoryServer`, to represent a Kopia repository server.
+It is a namespace-scoped resource, owned and managed by the controller in the
+`kanister` namespace. As mentioned above, the CRD offers a set of parameters to
+configure a Kopia repository server.
+
+To limit the controller's RBAC scope, all the `RepositoryServer` resources
+are created in the `kanister` namespace.
+
+A sample `RepositoryServer` resource created to interact with the Kopia
+repository of workloads running in the `monitoring` namespace looks like this:
+
+```yaml
+apiVersion: cr.kanister.io/v1alpha1
+kind: RepositoryServer
+metadata:
+   name: repository-monitoring
+   namespace: kanister
+   labels:
+      repo.kanister.io/target-namespace: monitoring
+spec:
+   storage:
+      # required: storage location info
+      secretRef:
+         name: location
+         namespace: kanister
+      # required: creds to access the location
+      # optional when using location type file-store
+      credentialSecretRef:
+         name: loc-creds
+         namespace: kanister
```

Comment:
Is it safe to assume that these secrets are always going to be in the kanister (controller's) namespace. If yes, do we need the namespace field?

Reply:
We have discussed this over a call with @pavannd1. This was intentionally added because we want the CR-controller to allow fetching secrets user create in their own namespace. This would be in-sync with the discussions to allow Blueprints or Profile CRs to be created in user/app namespace and not the controller(kanister) namespace